### PR TITLE
Document `Res` and its friends

### PR DESCRIPTION
**Pull request by camelid**
I noticed [this Zulip conversation][z] and thought it would be a good idea to
document `Res` and the other types near it.

[z]: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.2382516.20-.20Add.20inherent.20associated.20types/near/227914819

### CODE REVIEW

File: compiler/rustc_hir/src/def.rs [hunk 1]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+/// - the call to `str_to_string` will resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to the definition of `str_to_string` in the current crate.
+//
 #[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
 #[derive(HashStable_Generic)]
 pub enum Res<Id = hir::HirId> {
+    /// Any ordinary definition not covered by the other `Res` variants.
+    ///
+    /// **Not bound to a specific namespace.**
     Def(DefKind, DefId),
```
Comment (camelid) [inline compiler/rustc_hir/src/def.rs:241]: Seems like this should be moved down to where `Err` is or `Err` should be moved up to here since they are both general to any namespace.

File: compiler/rustc_hir/src/def.rs [hunk 2]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+    /// would have the [`DefId`] of the impl associated with it. Finally, the `Self` in
+    ///
+    /// ```
+    /// impl Foo for Bar {
+    ///     fn foo() -> Box<Self> { Box::new(Bar) }
+    /// }
+    /// ```
+    ///
+    /// would have both the [`DefId`] of `Foo` and the [`DefId`] of the impl
+    /// associated with it.
```
Comment (camelid) [inline compiler/rustc_hir/src/def.rs:280]: Is this description accurate?

File: compiler/rustc_hir/src/def.rs [hunk 3]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+    ),
+    /// A tool attribute module; e.g., the `rustfmt` in `#[rustfmt::skip]`.
+    ///
+    /// **Belongs to the type namespace.**
+    ToolMod,
 
     // Value namespace
-    SelfCtor(DefId /* impl */), // `DefId` refers to the impl
+    /// The `Self` constructor, along with the [`DefId`]
+    /// of the impl it is associated with.
```
Comment (camelid) [inline compiler/rustc_hir/src/def.rs:308]: Is this description accurate?

File: compiler/rustc_hir/src/def.rs [hunk 4]
```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
...
+    /// Type parameter: the `T` in `Vec<T>`
     TyParam,
 
     // Value namespace
     Fn,
     Const,
+    /// Const-generic type parameter.
     ConstParam,
     Static,
     /// Refers to the struct or enum variant's constructor.
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:81]: It might help to give an example of when the Ctor is different from the type itself: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=af764b9af1c0b2373e9f8a28484231c2

File: compiler/rustc_hir/src/def.rs [hunk 5]
```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
...
     Union,
     Enum,
-    /// Refers to the variant itself, `DefKind::Ctor` refers to its constructor if it exists.
+    /// Refers to the variant itself, [`DefKind::Ctor`] refers to its constructor if it exists.
     Variant,
     Trait,
-    /// `type Foo = Bar;`
+    /// Type alias: `type Foo = Bar;`
     TyAlias,
     ForeignTy,
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:67]:
```suggestion
    /// Type from an `extern` block.
    ForeignTy,
```

File: compiler/rustc_hir/src/def.rs [hunk 6]
```diff
@@ -83,10 +92,14 @@ pub enum DefKind {
     ExternCrate,
     Use,
     ForeignMod,
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:94]:
```suggestion
    /// An `extern` block.
    ForeignMod,
```

File: compiler/rustc_hir/src/def.rs (hunk 2 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:280]: I think so, yes: https://github.com/rust-lang/rust/blob/861872bc453bde79b83ff99d443d035225f10e87/compiler/rustc_resolve/src/late.rs#L1265

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:308]: No, constructors are not associated with an impl block. In the playground link I gave above, the `Y` that isn't re-exported to the crate root is the Ctor. I'm not sure if the DefId here is the id of the Ctor or of the type itself, petrochenkov may know.

File: compiler/rustc_hir/src/def.rs [hunk 7]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+    /// **Belongs to the value namespace.**
+    ///
+    /// *See also [`Res::SelfTy`].*
+    SelfCtor(DefId),
+    /// A local variable or function parameter.
+    ///
+    /// **Belongs to the value namespace.**
     Local(Id),
 
     // Macro namespace
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:319]:
```suggestion
```

File: compiler/rustc_hir/src/def.rs [hunk 8]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+    /// An attribute that is *not* implemented via macro.
+    /// E.g., `#[inline]` and `#[rustfmt::skip]`, which are essentially directives,
+    /// as opposed to `#[test]`, which is a builtin macro.
+    ///
+    /// **Belongs to the macro namespace.**
     NonMacroAttr(NonMacroAttrKind), // e.g., `#[inline]` or `#[rustfmt::skip]`
 
     // All namespaces
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:327]:
```suggestion
```

File: compiler/rustc_hir/src/def.rs [hunk 9]
```diff
@@ -275,17 +379,23 @@ impl PartialRes {
...
-///
-/// Therefore, they have a separate universe (namespace).
+/// Different kinds of symbols can coexist even if they share the same textual name.
+/// Therefore, they each have a separate universe (known as a "namespace").
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
 pub enum Namespace {
+    /// The type namespace includes `struct`s, `enum`s, `union`s, `trait`s, and `mod`s
+    /// (and, by extension, crates).
     TypeNS,
+    /// The value namespace includes `fn`s, `const`s, and `static`s.
```
Comment (jyn514) [inline compiler/rustc_hir/src/def.rs:389]:
```suggestion
    /// The value namespace includes `fn`s, `const`s, `static`s, and local variables.
```

File: compiler/rustc_hir/src/def.rs [hunk 10]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+//
 #[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
 #[derive(HashStable_Generic)]
 pub enum Res<Id = hir::HirId> {
+    /// Any ordinary definition not covered by the other `Res` variants.
+    ///
+    /// **Not bound to a specific namespace.**
     Def(DefKind, DefId),
 
     // Type namespace
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:243]:
```suggestion
```

File: compiler/rustc_hir/src/def.rs [hunk 11]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+    /// ```
+    ///
+    /// would have both the [`DefId`] of `Foo` and the [`DefId`] of the impl
+    /// associated with it.
+    ///
+    /// *See also [`Res::SelfCtor`].*
+    ///
+    /// -----
     ///
     /// HACK(min_const_generics): impl self types also have an optional requirement to not mention
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:286]:
```suggestion
    /// HACK(min_const_generics): impl self types also have an optional requirement to **not** mention
```

File: compiler/rustc_hir/src/def.rs [hunk 12]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
     /// any generic parameters to allow the following with `min_const_generics`:
-    /// ```rust
+    /// ```
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:288]: why did you remove the "rust" here?

File: compiler/rustc_hir/src/def.rs [hunk 13]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
     /// impl Foo { fn test() -> [u8; std::mem::size_of::<Self>()] {} }
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:289]:
```suggestion
    /// impl Foo { fn test() -> [u8; std::mem::size_of::<Self>()] { todo!() } }
```
:sparkles:

File: compiler/rustc_hir/src/def.rs [hunk 14]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+        Option<DefId>,
+        /// Optionally, the impl associated with this `Self` type.
+        Option<(DefId, bool)>,
+    ),
+    /// A tool attribute module; e.g., the `rustfmt` in `#[rustfmt::skip]`.
+    ///
+    /// **Belongs to the type namespace.**
+    ToolMod,
 
     // Value namespace
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:306]:
```suggestion
```

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (lcnr) [reply compiler/rustc_hir/src/def.rs:308]: yeah, the constructor should contain the `DefId` of either the struct or the variant it is associated with i think.

They aren't really associated with impls

File: compiler/rustc_hir/src/def.rs [hunk 15]
```diff
@@ -275,17 +379,23 @@ impl PartialRes {
...
 
-/// Different kinds of symbols don't influence each other.
-///
-/// Therefore, they have a separate universe (namespace).
+/// Different kinds of symbols can coexist even if they share the same textual name.
+/// Therefore, they each have a separate universe (known as a "namespace").
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
 pub enum Namespace {
+    /// The type namespace includes `struct`s, `enum`s, `union`s, `trait`s, and `mod`s
+    /// (and, by extension, crates).
```
Comment (lcnr) [inline compiler/rustc_hir/src/def.rs:387]: apparently also tool attribute modules if the comment for that res is correct. Not sure how sensible it is to exhaustively enumerate everything that's possible here.

Might be good to put an "e.g." in front here

**Review by lcnr — COMMENTED** [review]
a bunch of nits and suggestions

i think these comments are really helpful, thanks :heart:

File: compiler/rustc_hir/src/def.rs (hunk 12 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:288]: Rustdoc adds it by default.

File: compiler/rustc_hir/src/def.rs [hunk 16]
```diff
@@ -31,6 +32,7 @@ pub enum CtorKind {
     Fictive,
 }
 
+/// An attribute that is not a macro; e.g., `#[inline]` or `#[rustfmt::skip]`.
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:35]: Is it really helpful to add tautological comments like this just for the sake of it?
The name tells what it is, and comments on variants already describe the individual cases.

File: compiler/rustc_hir/src/def.rs [hunk 17]
```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
...
     Enum,
-    /// Refers to the variant itself, `DefKind::Ctor` refers to its constructor if it exists.
+    /// Refers to the variant itself, [`DefKind::Ctor`] refers to its constructor if it exists.
     Variant,
     Trait,
-    /// `type Foo = Bar;`
+    /// Type alias: `type Foo = Bar;`
     TyAlias,
     ForeignTy,
+    /// Trait alias: `type IntIterator = Iterator<Item = i32>;`
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:68]: Wrong example, not a trait alias.

File: compiler/rustc_hir/src/def.rs [hunk 18]
```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
...
     TraitAlias,
+    /// Associated type: `trait MyTrait { type Assoc; }`
     AssocTy,
+    /// Type parameter: the `T` in `Vec<T>`
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:72]:
```suggestion
    /// Type parameter: the `T` in `struct Vec<T> { ... }`
```
We are talking about definitions of type parameters here, not about use of type arguments.

File: compiler/rustc_hir/src/def.rs [hunk 19]
```diff
@@ -47,33 +49,40 @@ pub enum NonMacroAttrKind {
...
     TyParam,
 
     // Value namespace
     Fn,
     Const,
+    /// Const-generic type parameter.
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:78]:
```suggestion
    /// Constant generic parameter `struct Foo<const S: usize> { ... }`.
```

File: compiler/rustc_hir/src/def.rs [hunk 20]
```diff
@@ -83,10 +92,14 @@ pub enum DefKind {
     ExternCrate,
     Use,
     ForeignMod,
+    /// Anonymous `const`, e.g. `const { 1 + 2}`
     AnonConst,
+    /// Opaque type, aka `impl Trait`.
     OpaqueTy,
     Field,
+    /// Lifetime parameter: the `'a` in `&'a str`
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:100]: Same mistake as with type parameters.

File: compiler/rustc_hir/src/def.rs [hunk 21]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+/// - `String::from` will also resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to `String::from`;
+/// - `s` will resolve to [`Res::Local`];
+/// - the call to `str_to_string` will resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to the definition of `str_to_string` in the current crate.
+//
 #[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
 #[derive(HashStable_Generic)]
 pub enum Res<Id = hir::HirId> {
+    /// Any ordinary definition not covered by the other `Res` variants.
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:238]:
```suggestion
    /// Definition having a unique ID (`DefId`), corresponds to something defined in user code.
```

File: compiler/rustc_hir/src/def.rs (hunk 16 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:35]: I did this for two reasons:

1. We use a lot of abbreviations in rustc that can be confusing to newcomers.
2. I wanted to include an example of this kind of attribute, but having the entirety of the docs be the word `#[inline]` is pretty confusing.

File: compiler/rustc_hir/src/def.rs (hunk 5 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:67]: I didn't document this one because IIRC `ForeignTy` means things other than `extern type`. But maybe that's just in rustdoc.

File: compiler/rustc_hir/src/def.rs (hunk 17 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:68]: Ah, it should be this instead, right?

```suggestion
    /// Trait alias: `trait IntIterator = Iterator<Item = i32>;`
```

**Comment by crlf0710** [comment]
@camelid Ping from triage! Any updates on this?

**Comment by camelid** [comment]
Yeah, it's on my todo list, I just need to sit down and address all the comments :)

File: compiler/rustc_hir/src/def.rs (hunk 19 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:78]: (replaced `S` with `N`)

```suggestion
    /// Constant generic parameter `struct Foo<const N: usize> { ... }`.
```

File: compiler/rustc_hir/src/def.rs [hunk 22]
```diff
@@ -83,10 +92,14 @@ pub enum DefKind {
     ExternCrate,
     Use,
     ForeignMod,
+    /// Anonymous `const`, e.g. `const { 1 + 2}`
```
[the remark this answers is not in the document]
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:95]: (combined examples)

```suggestion
    /// Anonymous constant, e.g. `1 + 2` in `[u8; 1 + 2]`, or `const { 1 + 2}`
```

File: compiler/rustc_hir/src/def.rs (hunk 20 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:100]:
```suggestion
    /// Lifetime parameter: the `'a` in `struct Foo<'a> { ... }`
```

File: compiler/rustc_hir/src/def.rs (hunk 9 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:389]: (added a note to clarify things a bit)

```suggestion
    /// The value namespace includes `fn`s, `const`s, `static`s, and local variables (including function arguments).
```

File: compiler/rustc_hir/src/def.rs [hunk 23]
```diff
@@ -196,35 +209,126 @@ impl DefKind {
...
+///
+/// - `str` will resolve to [`Res::PrimTy`];
+/// - `String` will resolve to [`Res::Def`], and the `Res` will include the [`DefId`]
+///   for `String` as defined in the standard library;
+/// - `String::from` will also resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to `String::from`;
+/// - `s` will resolve to [`Res::Local`];
+/// - the call to `str_to_string` will resolve to [`Res::Def`], with the [`DefId`]
+///   pointing to the definition of `str_to_string` in the current crate.
+//
```
[the remark this answers is not in the document]
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:236]: How's this? e601da2eea3

File: compiler/rustc_hir/src/def.rs (hunk 10 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:249]: I think these have some value and I don't want to remove them in this PR.

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:314]: Hmm, I grepped through the code and looked around a bit, and it seems that the `DefId` *is* from an `impl` block: https://github.com/rust-lang/rust/blob/8e6b478a5a146ad77ae4340ce379e1674c35ad69/compiler/rustc_resolve/src/late.rs#L1277

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:82]: I don't think I understand the example well enough to document this; perhaps you or someone else could add some documentation in a follow-up? I would like to land this PR and then we can do follow-ups.

**Comment by camelid** [comment]
I think this is almost ready! Unresolved reviews:

- [ ] https://github.com/rust-lang/rust/pull/82958#discussion_r590992695
- [ ] https://github.com/rust-lang/rust/pull/82958#discussion_r591094295
- [ ] https://github.com/rust-lang/rust/pull/82958#discussion_r591630843
- [ ] https://github.com/rust-lang/rust/pull/82958#discussion_r591843999

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:82]: Sure, seems good. I came up with a slightly more extended example in the meantime so I don't forget https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ed6253f734e68b2594bb6715886fa019 - tl;dr SelfCtor is in the value namespace but the type is in the type namespace.

File: compiler/rustc_hir/src/def.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (camsteffen) [reply compiler/rustc_hir/src/def.rs:240]: Great!

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:314]: That's not what that code is doing - it's adding `Self` to the value namespace for items within that impl block, which is what allows things like
```rust
struct S {}
impl S {
  fn f() -> S { ... }
}
```
But SelfCtor isn't limited to impl blocks, `struct S;` also generates a Ctor named `S`.

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:82]: Oh, I can say something about how they're in different namespaces—is that what you meant?

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:314]: `struct S;` generates a `Res::Def(DefKind::Ctor)` but IIUC `Res::SelfCtor` is **only** for `impl` blocks. The docs this thread is about are for `SelfCtor`.

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:82]: Yes, that's what I meant :)

Comment (camelid) [reply compiler/rustc_hir/src/def.rs:82]: Hmm, now I feel confused: why do both `Variant` and `Ctor` exist? `Variant`s are in the value namespace, aren't they?

File: compiler/rustc_hir/src/def.rs (hunk 3 above)
Comment (jyn514) [reply compiler/rustc_hir/src/def.rs:314]: Ah ok, I didn't realize they were different. This seems fine then.

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (petrochenkov) [reply compiler/rustc_hir/src/def.rs:82]: Variants are in type namespace, https://github.com/rust-lang/rfcs/pull/2593 can give more intuition on why is this the case.
Variant constructors are values - callable (for tuple constructors) or ~not (for unit constructors).

File: compiler/rustc_hir/src/def.rs (hunk 16 above)
Comment (petrochenkov) [reply compiler/rustc_hir/src/def.rs:35]: I guess it's a question of who these comments are optimized for, this *may* be useful for a person looking at this enum for the first time, but for anyone else using this as a reference/reminder (which arguably happens more often) tautologies and walls of text are distractions and a nuisance.

File: compiler/rustc_hir/src/def.rs (hunk 4 above)
Comment (camelid) [reply compiler/rustc_hir/src/def.rs:82]: Ah, I was wondering if that RFC fit into this! I added an explanation based on what you described. 1e05d9e7753

**Comment by camelid** [comment]
Ok, all the reviews have been resolved! 🎉 

Let me know if you want me to squash and then this should be ready to merge.

**Comment by petrochenkov** [comment]
r=me with commits squashed.

**Comment by camelid** [comment]
(removed an extraneous period in the docs)
